### PR TITLE
Allow modifying macro tokens at parse time

### PR DESCRIPTION
**Pull request by igelado**
This pull requests adds a callback to ParseCallbacks that enables users to mangle the tokens on a macro definition. The goal is to allow custom "sanitization" of macros to enable generation of rust code.

In my use case, I have a few header files that define (a few hundred) macros such as:

```
#define FLAG_BITS            8:10
#define FLAG_FOO            0x1
#define FLAG_BAR            0x2
#define FLAG_FOOBAR    0x4
```

Currently, bindgen, rightly, ignores `FLAG_BITS` since the meaning of that definition is application specific. In my case, the side effect is that I lose the information about how much to shift bits around.

This pull request allows programmers to define a `modify_macro` callback that enables inspecting the Tokens of a given macro definition and sanitize it for bindgen to be able to generate rust code for it.

In my use case, the callback translates the three tokens `8` `:` `10` into `(8 << 16) | 10`. A value I can use in my rust code to determine the location of the bits.

### CODE REVIEW

File: src/ir/var.rs [hunk 1]
```diff
@@ -220,7 +220,7 @@ impl ClangSubItemParser for Var {
                     handle_function_macro(&cursor, &tokens, callbacks)?;
                 }
 
-                let value = parse_macro(ctx, &tokens);
+                let value = parse_macro(ctx, &cursor.spelling(), &tokens);
```
Comment (emilio) [inline src/ir/var.rs:223]: It seems odd to pass the name via `cursor.spelling()` if we're not using that as the real name (in fact `modify_macro` could actually change the macro name...). But anyway.

**Review by emilio — CHANGES_REQUESTED** [review]
Can you elaborate with a concrete example for this feature? I'm a bit skeptic that exposing raw cexpr tokens is the right abstraction...

Implementation wise looks reasonable though... but needs tests. Can you add some to `bindgen-integration`? You should be able to cargo-cult from the other macro callback tests.

**Comment by emilio** [comment]
Err, sorry, should've read the PR message more carefully. Rather than modifying the tokens, would it be reasonable to pass the tokens to `will_parse_macro` (without modifying them)?

**Comment by igelado** [comment]
Passing tokens to `will_parse_macro` could help to identify problematic macros, but I believe that no other callback for that macro will be ever executed since the call to clang parse will fail (the macro is neither `int_macro` nor `str_macro`). Without modifying the tokens to make clang happy I fail to see how this will work. What do you have in mind?

File: src/ir/var.rs (hunk 1 above)
Comment (igelado) [reply src/ir/var.rs:223]: Yeah! Probably the interface would be better if we pass the name as a String.

For the tokens, we could pass the slice that omits the macro name; that would work for my use case. However my line of thinking in passing the mutable reference to the tokens is that as part of the  process we could want to change the name. For instance, another reasonable approach for my use case is to transform
 ```
#define FLAG_BITS 8:10
```
into
```
#define FLAG_HI_BIT 10
#define FLAG_LO_BIT 8
```

**Pull request closed without merging**